app/src/main/java/org/mozilla/focus/search/SearchEngine.java

```diff
@@ -101,4 +130,79 @@ private String paramSubstitution(String template, String query) {
 
         return template;
     }
+
+    public static boolean addSearchEngine(SharedPreferences sharedPreferences, Context context, String engineName, String searchQuery) {
+        final Bitmap iconBitmap = IconGenerator.generateLauncherIcon(context, searchQuery);
+        final String searchEngineXml = buildSearchEngineXML(engineName, searchQuery, iconBitmap);
+        if (searchEngineXml == null) {
+            return false;
+        }
+        final Set<String> existingEngines = sharedPreferences.getStringSet(PREF_KEY_CUSTOM_SEARCH_ENGINES, Collections.<String>emptySet());
+        final Set<String> newEngines = new LinkedHashSet<>();
+        newEngines.addAll(existingEngines);
+        newEngines.add(engineName);
+
+        sharedPreferences.edit().putInt(PREF_KEY_CUSTOM_SEARCH_VERSION, CUSTOM_SEARCH_VERSION)
+                .putStringSet(PREF_KEY_CUSTOM_SEARCH_ENGINES, newEngines)
+                .putString(engineName, searchEngineXml)
+                .apply();
+
+        // Force SearchEngineManager to refetch, to get the newest search engine.
+        SearchEngineManager.getInstance().init(context);
+        return true;
+    }
+
+    public static String buildSearchEngineXML(String engineName, String searchQuery, Bitmap iconBitmap) {
+        Document document = null;
+        try {
+            document = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument();
```

Comment:
I have been using `XmlSerializer` for writing XML to disk in the past. The advantage is that it can write to a stream immediately without the need to build up the document in memory first. But then again those files are rather small anyways.

Reply:
Cool, it looks like that's nice because I wouldn't need to do the conversion with this transformer factory (which seemed weirdly complicated, but I assumed it was just "because Java"). I won't do it here but I'll file it in a follow-up with the unit test.